content/posts/unpublished/working-with-test-containers/post.md

```diff
@@ -0,0 +1,167 @@
+# Working with test containers
+
+
+Framework that simplifies the process of setting up, managing, and tearing down containerized environments for integration testing.
+
+
+![](assets/testcontainers.png)
+
+
+Authors: Yauheni Navosha
+Date: unpublished
+Category: backend
+
+
+tags: backend, tests, testing, integration testing, docker, containers
+
+---
+
+## What is Testcontainers?
+
+Testcontainers is a testing library integration tests with real services wrapped in Docker containers. Using Testcontainers, you can write tests by directly interacting with the same kind of services utilized in production, without relying on mocks or in-memory services.
+
+## When to use it?
+
+Testcontainers simplifies running of integration tests that involve external dependencies such as databases, message queues, etc., by providing a convenient way to spin up disposable instances of these dependencies within you test environment.
+
+## Why should you use testcontainers instead of mocked and in-memory services?
```

Comment:
It might be worth to mention about the disadvantages(even if they are minor)?
ie: bringing another dependency,  needs to be run at least one so it might be time consuming(more or less). slower comparing to mock.

Reply:
IMO there is no sense for it. @tarun-bb what do you think?
@nafelaraz What do you mean by "slower comparing to mock?"

Reply:
I havent test it myself but I saw in multiple sources:
-You need to run containers at least once - it consumes time and resources. For example, PostgreSQL as a Docker image needs around 4 seconds to start on my machine, whereas the H2 in-memory database needs only 0.4 seconds. From my experience, Localstack which emulates AWS components, can start much longer, even 20 seconds on my machine.


-There are of course drawbacks to using TestContainers. Test cases will be much slower compared to H2. There are ways to mitigate this. The @Rule annotation brings up a new database for every test case in your class. Bringing up a database is a costly operation, and we can optimize this by using @ClassRule. When you use @ClassRule, one database is brought up for all the tests in the class. In this scenario, it becomes important to ensure that you are cleaning up the data in the database after a test case runs, to ensure test isolation. This can be done by using the @After lifecycle.

https://engineering.zalando.com/posts/2021/02/integration-tests-with-testcontainers.html
https://medium.com/miq-tech-and-analytics/testcontainers-the-modern-way-of-writing-database-tests-ed49554856e5

Reply:
Mentioning a few potential drawbacks of using test containers is a good point.

Reply:
updated

Reply:
I have not quite widely used the testcontainers, could you list scenarios where using in-memory databases or mocks is preferable?

Reply:
You can mention something like using `Test Containers increases the build time and takes a lot of resources. on the other side in memory mocking is the opposite.`

Reply:
@eltntawy It is already has been added https://github.com/Backbase/backbase.github.io/blob/4a9067545aa01346cb36e0a8ebb08eac14337f3d/content/posts/unpublished/working-with-test-containers/post.md?plain=1#L32